lib/compatibility.bzl

```diff
@@ -0,0 +1,130 @@
+            "@platforms//os:linux",
+            "@platforms//os:macos",
+        ),
+    )
+    ```
+
+    Args:
+      *settings: The `config_setting` or `constraint_value` targets.
+
+    Returns:
+      A native `select()` which maps any of the settings to the incompatible target.
+    """
+    return selects.with_or({
+        "//conditions:default": [],
+        tuple(settings): ["@bazel_skylib//lib/compatibility:none_of"],
+    })
+
+def _any_of(*settings):
+    """Create a `select()` for `target_compatible_with` which matches any of the given settings.
+
+    Any of the settings will resolve to an empty list, while the default condition will map to
+    an incompatible `constraint_value` for the purpose of target skipping.
+
+    In other words, use this function to make target incompatible unless one or more of the
+    settings are true.
+
+    ```python
+    cc_binary(
+        name = "bin",
+        srcs = ["bin.cc"],
+        # This target can only be built for Linux or Mac.
+        target_compatible_with = compatibility.any_of(
+            "@platforms//os:linux",
+            "@platforms//os:macos",
+        ),
+    )
+    ```
+
+    Args:
+      *settings: The `config_settings` or `constraint_value` targets.
+
+    Returns:
+      A native `select()` which maps any of the settings an empty list.
+    """
+    return selects.with_or({
+        tuple(settings): [],
+        "//conditions:default": ["@bazel_skylib//lib/compatibility:any_of"],
+    })
+
+def _all_of(*settings):
+    """Create a `select()` for `target_compatible_with` which matches all of the given settings.
+
+    All of the settings must be true to get an empty list. Failure to match will result
+    in an incompatible `constraint_value` for the purpose of target skipping.
+
+    In other words, use this function to make a target incompatible unless all of the settings are
+    true.
+
+    Example:
+
+    ```python
+    config_setting(
+        name = "dbg",
+        values = {"compilation_mode": "dbg"},
+    )
+
+    cc_binary(
+        name = "bin",
+        srcs = ["bin.cc"],
+        # This target can only be built for Linux in debug mode.
+        target_compatible_with = compatibility.all_of(
+            ":dbg",
+            "@platforms//os:linux",
+        ),
+    )
+    ```
+
+    See also: `selects.config_setting_group(match_all)`
+
+    Args:
+      *settings: The `config_setting` or `constraint_value` targets.
+
+    Returns:
+      A native series of `select()`s. The result is "incompatible" unless all settings are true.
+    """
+    if len(settings) > MAX_NUM_ALL_OF_SETTINGS:
+        fail("Cannot support more than {} arguments. Use selects.config_setting_group() instead.".format(MAX_NUM_ALL_OF_SETTINGS))
+
+    result = []
+    for i, setting in enumerate(settings):
+        result += select({
+            setting: [],
+            "//conditions:default": ["@bazel_skylib//lib/compatibility:all_of_{}".format(i)],
```

Comment:
What happens if you replace `"@bazel_skylib//lib/compatibility:all_of_{}".format(i)` with `"@platforms//:incompatible"` in this line?

It if works, it's much simpler.

Reply:
Same issue as https://github.com/bazelbuild/bazel-skylib/pull/381#discussion_r954544755

Reply:
There's one alternative, which I considered worse until now. How about:
```
result = []
for setting in enumerate(settings):
   result += select({
      setting: result,
       "//conditions:default": ["@bazel_skylib//lib/compatibility:incompatible_in_all_of"]
```

I think this should evaluate to a single label when a condition is not met.

First line could also be `result = settings`. That might work better in `cquery`-es.